### PR TITLE
os/bluestore: Add CoDel to BlueStore for Bufferbloat mitigation

### PR DESCRIPTION
**Pull request by esmaeil-mirvakili**
The CoDel algorithm has been implemented in the BlueStore to mitigate bufferbloat in the BlueStore backend. The CoDel algorithm limits the number of transactions submitted to BlueStore. The number of transactions is defined using a dual-loop optimization mechanism. The inner (fast) loop minimize the latency bu limiting submitted transaction, and the outer (slow) loop control the inner loop to achieve a tradeoff between throughput and latency.

Signed-off-by: Esmaeil Mirvakili <smirvaki@ucsc.edu>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by esmaeil-mirvakili** [comment]
@athanatos

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -16130,6 +16161,226 @@ void BlueStore::BlueStoreThrottle::complete(TransContext &txc)
 }
 #endif
 
+BlueStore::BlueStoreSlowFastCoDel::BlueStoreSlowFastCoDel(CephContext *_cct) :
+  fast_timer(_cct, fast_timer_lock), slow_timer(_cct, slow_timer_lock) {}
+
+BlueStore::BlueStoreSlowFastCoDel::~BlueStoreSlowFastCoDel() {
+  std::lock_guard l1{fast_timer_lock};
+  fast_timer.cancel_all_events();
+  fast_timer.shutdown();
```
Comment (athanatos) [inline src/os/bluestore/BlueStore.cc:16170]: These should probably be scoped so that fast_timer_lock is released prior to taking slow_timer_lock?

**Comment by athanatos** [comment]
I'll get into this properly tomorrow.

File: src/test/common/test_regression_utils.cc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  // the inverse matrix should have the same size
+  ASSERT_EQ(random_square_m_inv.size1(), random_square_m.size1());
+  ASSERT_EQ(random_square_m_inv.size2(), random_square_m.size2());
+  matrix<double> matrix_prod = prod(random_square_m, random_square_m_inv);
+  // the product should be an identity matrix
+  for ( int i = 0; i < matrix_prod.size1(); i++)
+    for (int j = 0; j < matrix_prod.size2(); j++)
+      if(i == j) {
+        ASSERT_TRUE(is_almost_equal(matrix_prod(i, j), 1, 1e-9));    // i == j -> 1
+      }else {
```
Comment (athanatos) [inline src/test/common/test_regression_utils.cc:61]: } else {

File: src/os/bluestore/BlueStore.h [hunk 3]
```diff
@@ -1826,8 +1828,97 @@ class BlueStore : public ObjectStore,
...
+    void reset(CephContext *cct);
+
+    void set_throttle(BlueStoreThrottle *_throttle);
+
+    void submit_txc_info(TransContext *txc);
+
+    uint64_t get_bluestore_budget();
+
+    bool is_activated();
+  protected:
```
Comment (athanatos) [inline src/os/bluestore/BlueStore.h:1857]: Reformat so that these lines are no longer than 80 characters.

File: src/os/bluestore/BlueStore.h [hunk 4]
```diff
@@ -1826,8 +1828,97 @@ class BlueStore : public ObjectStore,
...
+    bool activated = false;    // true if SlowFastCodel is activated
+    int64_t initial_fast_interval;     // Initial interval for fast loop (original codel)
```
Comment (athanatos) [inline src/os/bluestore/BlueStore.h:1859]: This and several other variables do not seem to get initialized either here or in the constructor colon initializer list.  Every one of these members must be initialized.

File: src/os/bluestore/BlueStore.h [hunk 5]
```diff
@@ -1826,8 +1828,97 @@ class BlueStore : public ObjectStore,
...
+    }
+    void reset_kv_throttle_max(int64_t m) {
+      throttle_bytes.reset_max(m);
+    }
   } throttle;
 
+  /**
+   * CoDel algorithm to calculate the incoming batch size to kv by using Throttle mechanism
+   */
+  class BlueStoreSlowFastCoDel {
```
Comment (athanatos) [inline src/os/bluestore/BlueStore.h:1842]: I don't think this needs to be embedded directly into BlueStore.  There seem to be two touchpoints between BlueStoreSlowFastCoDel and BlueStore's internal data structures.

1) submit_txc_info uses a TransContext to get two pieces of information: txc_state_proc_start and bytes.  Explicitely pass in bytes and operation latency instead, as well as the value of bs_throttle->get_kv_throttle_current().
2) bs_throttle only seems to be in one other place: reset_kv_throttle_max.  BlueStoreSlowFastCoDel could simply take a std::function<void(uint64_t)> argument in the constructor.

With those changes, you should be able to move this whole structure into its own .h and .cc files without any references to bluestore internals.  This would also enable you to write a unit test for it independently of bluestore.

**Comment by esmaeil-mirvakili** [comment]
@athanatos
The changes have been applied. Please review the PR again.

**Comment by esmaeil-mirvakili** [comment]
jenkins test make check

**Comment by esmaeil-mirvakili** [comment]
jenkins test api

**Comment by esmaeil-mirvakili** [comment]
jenkins test make check arm64

**Comment by esmaeil-mirvakili** [comment]
jenkins test make check arm64

**Comment by athanatos** [comment]
Prefix the first line of the first commit with os/bluestore: and the first line of the second commit with test/os/bluestore:

**Comment by athanatos** [comment]
Probably need someone more familiar with bluestore to have a look -- @aclamk @ifed01 ?

**Comment by athanatos** [comment]
I think you may have picked up an extra two commits when you rebased?

**Comment by athanatos** [comment]
How about we also add a bluestore facet in teuthology for codel:

```
diff --git a/qa/objectstore_debug/bluestore-codel.yaml b/qa/objectstore_debug/bluestore-codel.yaml
new file mode 100644
index 00000000000..a098573565f
--- /dev/null
+++ b/qa/objectstore_debug/bluestore-codel.yaml
@@ -0,0 +1,43 @@
+overrides:
+  thrashosds:
+    bdev_inject_crash: 2
+    bdev_inject_crash_probability: .5
+  ceph:
+    fs: xfs
+    conf:
+      osd:
+        osd objectstore: bluestore
+        bluestore block size: 96636764160
+        debug bluestore: 20
+        debug bluefs: 20
+        debug rocksdb: 10
+        bluestore fsck on mount: true
+        bluestore codel: true
+        # lower the full ratios since we can fill up a 100gb osd so quickly
+        mon osd full ratio: .9
+        mon osd backfillfull_ratio: .85
+        mon osd nearfull ratio: .8
+        osd failsafe full ratio: .95
+# this doesn't work with failures bc the log writes are not atomic across the two backends
+#        bluestore bluefs env mirror: true
+        bdev enable discard: true
+        bdev async discard: true
+  ceph-deploy:
+    fs: xfs
+    bluestore: yes
+    conf:
+      osd:
+        osd objectstore: bluestore
+        bluestore block size: 96636764160
+        debug bluestore: 20
+        debug bluefs: 20
+        debug rocksdb: 10
+        bluestore fsck on mount: true
+        # lower the full ratios since we can fill up a 100gb osd so quickly
+        mon osd full ratio: .9
+        mon osd backfillfull_ratio: .85
+        mon osd nearfull ratio: .8
+        osd failsafe full ratio: .95
+        bdev enable discard: true
+        bdev async discard: true
```

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -4573,6 +4573,13 @@ BlueStore::BlueStore(CephContext *cct,
   _init_logger();
   cct->_conf.add_observer(this);
   set_cache_shards(1);
+  codel = new BlueStoreSlowFastCoDel(cct,
```
Comment (athanatos) [inline src/os/bluestore/BlueStore.cc:4576]: Currently, we're checking for activated in a few places in BlueStoreSlowFastCodel.cc, but the timer etc are still running even when inactivated.  I'd rather just avoid instantiating codel at all while it's disabled and check for null in BlueStore before accessing it.  That would make it easier for developers to prove whether it was enabled when debugging problems.

**Comment by aclamk** [comment]
@esmaeil-mirvakili 
I believe that commits
44c2d89341acd1466dbc185e384ed9b25195ca3c "cephfs'
b8ba4c63b15da339ce481849a5bdf7f9d3f33522 "scrub"
are not a part of proposed solution.

**Comment by aclamk** [comment]
Modelling of latency - throughput relation

My understanding is that SSDs and HDDs have two different latency/throughput models.
I will propose models that I think approximate reality well enough so we can model relation lat/thru.

**SSD**

Lets analyze latency and throughput for a case when **n** operations are grouped together into a batch.

| batch latency component | time | notes |
| --- | --- | --- |
| prepare | Tprep | Assuming more worker threads then n |
| seek | n * Tseek | Time needed by controller to setup / erase previous content | 
| write | n * avg_op * Twritespeed | Avg_op is an average of operation sizes |

*Tprep includes time to fetch metadata on object if necessary, allocate disc space, etc.*

Transferred data is `n * avg_op`.
Batch execution time is `Tprep + n * Tseek + n * avg_op * Twritespeed`

`Lat = A + n * B`
`Data = n * D`

It is easy to see here that `Thru = Data/Lat` will asymptotically increase to D/B.


**HDD**

Lets analyze latency and throughput for a case when **n** operations are grouped together into a batch.

| latency component | time | notes |
| --- | --- | --- |
| prepare | Tprep | Assuming more worker threads then n |
| seek | n * Trotational_wait | Expected wait time before disk appears below head |
| seek | (max_pos - min_pos) * Tradial_seek | Sum of time needed for head to jump between tracks |
| write | n * avg_op * Twritespeed | avg_op is an average of operation sizes |

Transferred data is `n * avg_op`.
Batch execution time is `Tprep + n * Trotational_wait + (max_pos - min_pos) * Tradial_seek + n * avg_op * Twritespeed`

Rationale for (max_pos - min_pos) * Tradial_seek:
We have multiple write ops in one batch. Disk scheduler will order writes to reduce head seeks.
Assumption here is that time required to perform head seek is linearly proportional to seek distance.
Therefore time required to seek through set of position is proportional to `(max_pos - min_pos)`.

I will assume that write access pattern is uniform and disc space is a range <0, 1>.
Then we can represent min_pos and max_pos as:
`min_pos = 1/(n+1)`
`max_pos = 1 - 1/(n+1)`

<details>
<summary>my calculations, scratchpad</summary>
Expected value of min_pos is:
EX(min_pos) = Integral(PDFmin_pos(x) * x, 0, 1) 

PDFmin_pos is a probability distribution function. 
PDFmin_pos can be obtained from cumulative distribution function: 

PDFmin_pos = dCDFmin_pos/dx 
Where

CDFmin_pos(x) = P{min_pos < x} = ~P{all pos > x} = 1 - (1-x)^n

PDFmin_pos = -n*(1-x)^(n-1) * -1 = n * (1-x)^(n-1)
EX(min_pos) = Integral( n * (1-x)^(n-1) * x, 0, 1) = 
“Wolfram alpha integrator”
= -(1-x)^n * (nx+1)/(n+1) on [0, 1] = 0 - -1*(1)/(n+1) = 1/(n+1)

I did not calculate max_pos, just assumed it must be 1 - min_pos.
</details>

Updated batch exec time = `Tprep + n * Trotational_wait + (1 - 2/(n+1)) * Tradial_seek + n * avg_op * Twritespeed`

`Lat = A + n * B + 1/n * C`
`Data = n * D`
So,
`Thru = n * D / (A + n * B + C/n)`

Now relation is much different - we now get the case that with increasing size of a batch, we can get significantly better throughput.

**Comment by aclamk** [comment]
**Proposal**

I think we will get much better quality of latency / throughput control if manipulate count of ops in a single batch.
We will get then a stream of samples of form:
(batch op count, batch data size) -> (execution time)

The algorithm will manipulate number of operations in a batch.

Adaptation to new conditions will be achieved by manipulating count by +/- 1 and observing resulted latency.
This approach basically skips step of full model reconstruction, and replaces it with constantly adapting to local optimum of latency/throughput.

New adaptation formula will just be some median/mean filtration of resulting local slope and adapting to new optimum.

File: doc/dev/bluestore-codel.rst [hunk 7]
```diff
@@ -0,0 +1,41 @@
+============================================
```
Comment (aclamk) [inline doc/dev/bluestore-codel.rst:1]: +1 for excellent rst

File: src/common/regression_utils.h [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+  /***
+  * Find a logarithmic function in form of "y = a + b * ln(x)" which fits
+  * the given points (x_values and y_values).
+  * @param std::vector<double> x_values, x values for sample points
+  * @param std::vector<double> y_values, y values for sample points
+  * @param double theta[2], holds the a and b as output (theta[0] = a and theta[1] = b)
+  */
+  static void logarithmic_regression(
+    const std::vector<double> &x_values,
```
Comment (aclamk) [inline src/common/regression_utils.h:47]: I think this is a linear regression, but applied on values that were logarithmed.
I think we should modify function name.

https://www.stat.purdue.edu/~boli/stat512/lectures/topic3.pdf

File: src/common/options/global.yaml.in [hunk 9]
```diff
@@ -6086,4 +6086,73 @@ options:
...
+  type: int
+  level: advanced
+  desc: initial target latency for SlowFastCodel in ns
+  default: 5000000
+  with_legacy: true
+- name: bluestore_codel_slow_interval
+  type: int
+  level: advanced
+  desc: the interval of slow loop in SlowFastCodel in ns (this parameter should be larger that 'bluestore_codel_fast_interval')
+  default: 500000000
```
Comment (aclamk) [inline src/common/options/global.yaml.in:6115]: How about converting time to seconds, or milliseconds but allow them to be floats?

**Comment by aclamk** [comment]
It feels to me like the learn-control loop of CoDel algorithm requires specific conditions between latency and throughput.

In learning step:
   we commit to fixed target latency and we measure obtained throughput.
In adaptation step:
   we expect that throughput can be manipulated by modifying target latency.
This works very good when BlueStore is saturated with IOs.

Those conditions will not hold in cases when BlueStore is not saturated with IOs.
When incoming ops are processed immediately they suffer only latency specific to BlueStore processing stages. 
When incoming traffic is low, ops mostly do not overlap and the execution latency stays within the noise threshold. As a result, for a wide range of target latency set we get exactly the same throughput. Throughput that will be exactly the same as incoming traffic.

It feels to me like we should suspend CoDel adaptation when BlueStore IO is not saturated.

**Comment by aclamk** [comment]
In most cases READ operations affect execution time of WRITE ops.
Fetching metadata for W op and flushing data to disk for W op share components/data with fetching metadata for R op and retrieving data for R op.

In read-heavy loads BlueStore will react differently to throttle control then in read-light loads.
One should expect that 75%R/25%W will react 3 times slower than 25%R/75%W for similar latency -> throughput manipulation.

Basically, I am worried that a model constructed when read load was low will poorly model conditions under heavy reads.

File: src/os/bluestore/BlueStore.cc [hunk 10]
```diff
@@ -4600,6 +4600,15 @@ BlueStore::BlueStore(CephContext *cct,
   _init_logger();
   cct->_conf.add_observer(this);
   set_cache_shards(1);
+  if ( cct->_conf->bluestore_codel) {
+    codel = new BlueStoreSlowFastCoDel(cct,
```
Comment (athanatos) [inline src/os/bluestore/BlueStore.cc:4604]: This is a memory leak since you don't release it in ~BlueStore.  Change the type of code to std::unique_ptr to avoid needing to release in ~BlueStore.

File: src/os/bluestore/BlueStoreSlowFastCoDel.h [hunk 11]
```diff
@@ -0,0 +1,132 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_OS_BLUESTORE_SLOWFASTCODEL_H
+#define CEPH_OS_BLUESTORE_SLOWFASTCODEL_H
```
Comment (athanatos) [inline src/os/bluestore/BlueStoreSlowFastCoDel.h:5]: Replace this and other guards with #pragma once

File: src/os/bluestore/BlueStore.cc (hunk 10 above)
Comment (athanatos) [reply src/os/bluestore/BlueStore.cc:4604]: Also, rewrap to

```
  if ( cct->_conf->bluestore_codel) {
    codel = new BlueStoreSlowFastCoDel(
      cct,
      [this](int64_t x) mutable {
	this->throttle.reset_kv_throttle_max(x);
      },
      [this]() mutable {
	return this->throttle.get_kv_throttle_current();
      });
  }
```

File: src/test/os/bluestore/TestBlueStoreSlowFastCoDel.cc [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+  static void SetUpTestCase() {}
+
+  static void TearDownTestCase() {}
+
+  void SetUp() override {
+    ceph_context = (new CephContext(CEPH_ENTITY_TYPE_ANY))->get();
+  }
+
+  void create_bluestore_slow_fast_codel() {
+    slow_fast_codel = new BlueStoreSlowFastCoDelMock(ceph_context,
```
Comment (athanatos) [inline src/test/os/bluestore/TestBlueStoreSlowFastCoDel.cc:122]: rewrap to 80 characters as I commented in BlueStore::BlueStore.

**Comment by athanatos** [comment]
@esmaeil-mirvakili Have @aclamk 's comments above been addressed?

**Comment by athanatos** [comment]
@esmaeil-mirvakili Is this ready for review?

**Comment by esmaeil-mirvakili** [comment]
@athanatos It is ready for review, but something is wrong with the checks. I'll run them again.

**Comment by esmaeil-mirvakili** [comment]
jenkins test docs

**Comment by esmaeil-mirvakili** [comment]
jenkins test make check

**Comment by athanatos** [comment]
@aclamk Can you take another look?  I'm inclined to get this merged sooner rather than later to make testing/iteration easier.  It's off by default and (imho) pretty well isolated from the rest of BlueStore, so it should be low risk.

**Review by aclamk — APPROVED** [review]
I do think now that we should merge this.
I do not think it is ready right now to enable and give to customer,
but it is an important framework that we must have to even be able to improve on actual adaptation algorithm.

**Comment by neha-ojha** [comment]
@athanatos @esmaeil-mirvakili seems like this PR is causing failures like https://jenkins.ceph.com/job/ceph-pull-requests/94227/consoleFull#14584389247725d52c-2930-43cb-b77a-ae0a919c2170. I suspect it is the same failure which was seen in https://trello.com/c/OmPlJ1pq/1513-wip-yuri4-testing-2022-04-22-0932#comment-626317e677cc78785021032d

```
[ RUN      ] TestSlowFastCoDel.test1
../src/common/Timer.cc: In function 'Context *CommonSafeTimer<ceph::mutex_debug_detail::mutex_debug_impl<false>>::add_event_after(ceph::timespan, Context *) [Mutex = ceph::mutex_debug_detail::mutex_debug_impl<false>]' thread 7f192d101300 time 2022-04-25T15:52:07.507067+0000
../src/common/Timer.cc: 132: FAILED ceph_assert(((lock).is_locked()))
 ceph version Development (no_version) quincy (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1b4) [0x7f192f31fb44]
 2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x140698f) [0x7f192f31f98f]
 3: (CommonSafeTimer<ceph::mutex_debug_detail::mutex_debug_impl<false> >::add_event_after(std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >, Context*)+0x47) [0x7f192f2e63d7]
 4: (BlueStoreSlowFastCoDel::_fast_interval_process()+0x1c5) [0x526765]
 5: (BlueStoreSlowFastCoDel::on_config_changed(ceph::common::CephContext*)+0x285) [0x525fc5]
 6: (BlueStoreSlowFastCoDel::BlueStoreSlowFastCoDel(ceph::common::CephContext*, std::function<void (long)>, std::function<long ()>)+0x22d) [0x525c4d]
 7: (BlueStoreSlowFastCoDelMock::BlueStoreSlowFastCoDelMock(ceph::common::CephContext*, std::function<void (long)>, std::function<long ()>, std::mutex&, std::condition_variable&, long, long, long, double)+0x94) [0x4b7884]
 8: (TestSlowFastCoDel::create_bluestore_slow_fast_codel()+0xd0) [0x4b28c0]
 9: (TestSlowFastCoDel_test1_Test::TestBody()+0x19) [0x4b19a9]
 10: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x7b) [0x51d7fb]
 11: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x6a) [0x501d0a]
 12: (testing::Test::Run()+0xc3) [0x4dfd73]
 13: (testing::TestInfo::Run()+0xda) [0x4e0aba]
 14: (testing::TestSuite::Run()+0xfb) [0x4e12db]
 15: (testing::internal::UnitTestImpl::RunAllTests()+0x442) [0x4ee9f2]
 16: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x7b) [0x52097b]
 17: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x6a) [0x5045ca]
 18: (testing::UnitTest::Run()+0xcb) [0x4ee55b]
 19: (RUN_ALL_TESTS()+0x11) [0x4bd691]
 20: main()
 21: __libc_start_main()
 22: _start()
```

**Comment by ljflores** [comment]
> @athanatos @esmaeil-mirvakili seems like this PR is causing failures like https://jenkins.ceph.com/job/ceph-pull-requests/94227/consoleFull#14584389247725d52c-2930-43cb-b77a-ae0a919c2170. I suspect it is the same failure which was seen in https://trello.com/c/OmPlJ1pq/1513-wip-yuri4-testing-2022-04-22-0932#comment-626317e677cc78785021032d
> 
> ```
> [ RUN      ] TestSlowFastCoDel.test1
> ../src/common/Timer.cc: In function 'Context *CommonSafeTimer<ceph::mutex_debug_detail::mutex_debug_impl<false>>::add_event_after(ceph::timespan, Context *) [Mutex = ceph::mutex_debug_detail::mutex_debug_impl<false>]' thread 7f192d101300 time 2022-04-25T15:52:07.507067+0000
> ../src/common/Timer.cc: 132: FAILED ceph_assert(((lock).is_locked()))
>  ceph version Development (no_version) quincy (dev)
>  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1b4) [0x7f192f31fb44]
>  2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x140698f) [0x7f192f31f98f]
>  3: (CommonSafeTimer<ceph::mutex_debug_detail::mutex_debug_impl<false> >::add_event_after(std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >, Context*)+0x47) [0x7f192f2e63d7]
>  4: (BlueStoreSlowFastCoDel::_fast_interval_process()+0x1c5) [0x526765]
>  5: (BlueStoreSlowFastCoDel::on_config_changed(ceph::common::CephContext*)+0x285) [0x525fc5]
>  6: (BlueStoreSlowFastCoDel::BlueStoreSlowFastCoDel(ceph::common::CephContext*, std::function<void (long)>, std::function<long ()>)+0x22d) [0x525c4d]
>  7: (BlueStoreSlowFastCoDelMock::BlueStoreSlowFastCoDelMock(ceph::common::CephContext*, std::function<void (long)>, std::function<long ()>, std::mutex&, std::condition_variable&, long, long, long, double)+0x94) [0x4b7884]
>  8: (TestSlowFastCoDel::create_bluestore_slow_fast_codel()+0xd0) [0x4b28c0]
>  9: (TestSlowFastCoDel_test1_Test::TestBody()+0x19) [0x4b19a9]
>  10: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x7b) [0x51d7fb]
>  11: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x6a) [0x501d0a]
>  12: (testing::Test::Run()+0xc3) [0x4dfd73]
>  13: (testing::TestInfo::Run()+0xda) [0x4e0aba]
>  14: (testing::TestSuite::Run()+0xfb) [0x4e12db]
>  15: (testing::internal::UnitTestImpl::RunAllTests()+0x442) [0x4ee9f2]
>  16: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x7b) [0x52097b]
>  17: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x6a) [0x5045ca]
>  18: (testing::UnitTest::Run()+0xcb) [0x4ee55b]
>  19: (RUN_ALL_TESTS()+0x11) [0x4bd691]
>  20: main()
>  21: __libc_start_main()
>  22: _start()
> ```

Tracker: https://tracker.ceph.com/issues/55433

**Comment by athanatos** [comment]
I'm going to go ahead and revert it for now.

**Comment by esmaeil-mirvakili** [comment]
I checked the error and the code, it seems that the problem is only the tests. For implementing the interval-based loops in codel, I used the SafeTimer similar to [https://github.com/ceph/ceph/blob/a9cb584d3f07ab30a69883f21b309e8072cc8237/src/test/common/test_lockdep.cc#L19](https://github.com/ceph/ceph/blob/a9cb584d3f07ab30a69883f21b309e8072cc8237/src/common/Throttle.cc#L875). However the same code is working in other parts of ceph, it can fail when CEPH_DEBUG_MUTEX is defined. When this macro is defined, the timer does some double-checks on mutex lock which will not work with this usage of timer.
One solution is to skip the test if this macro exists for now like this test: [https://github.com/ceph/ceph/blob/a9cb584d3f07ab30a69883f21b309e8072cc8237/src/test/common/test_lockdep.cc#L19](https://github.com/ceph/ceph/blob/a9cb584d3f07ab30a69883f21b309e8072cc8237/src/test/common/test_lockdep.cc#L19)

**Comment by athanatos** [comment]
@esmaeil-mirvakili Which double check is it failing?

**Comment by esmaeil-mirvakili** [comment]
It is checking that the timer mutex is locked when calling "add_event_after" [here](https://github.com/ceph/ceph/blob/a9cb584d3f07ab30a69883f21b309e8072cc8237/src/common/Timer.cc#L132).
And in "[ceph_mutex.h](https://github.com/ceph/ceph/blob/master/src/common/ceph_mutex.h)", if CEPH_DEBUG_MUTEX is not defined it returns true all the time, otherwise it check the lock.

We didn't have any problem with tests before, has something changed in tests or tests after merge are different?

**Comment by athanatos** [comment]
The reason for that check is that failing to lock the timer mutex when calling add_event_after causes elements of the safe_timer api to have race conditions.  Are you quite sure that your implementation is actually correct without holding the lock?

**Comment by athanatos** [comment]
@esmaeil-mirvakili Were you able to look into holding the lock while using add_event_after?
